hack/build/release/release.go

```diff
@@ -2,116 +2,357 @@ package release
+		}
+
+		if strings.HasPrefix(pre, "-rc") {
+			rc, err := strconv.ParseInt(pre[3:], 10, 64)
+			if err != nil {
+				return "", err
+			}
+
+			pre = fmt.Sprintf("-rc%d", rc+1)
+		} else {
+			pre = "-rc1"
+		}
+
+		return fmt.Sprintf("v%d.%d.%d%s", major, minor, patch, pre), nil
+	}
+
+	if pre != "" {
+		// latest release was a pre-release and so we can consider
+		// the next release just drops the "pre" part for any version.
+		return fmt.Sprintf("v%d.%d.%d", major, minor, patch), nil
+	}
 
-	tagList, err := base.WithExec([]string{"git", "tag", "--list", "v*"}).Stdout(ctx)
+	switch vp {
+	case PATCH:
+		patch++
+	case MINOR:
+		patch = 0
+		minor++
+	case MAJOR:
+		patch = 0
+		minor = 0
+		major++
+	default:
+		return "", fmt.Errorf("unexpected version part: %q", vp)
+	}
+
+	return fmt.Sprintf("v%d.%d.%d%s", major, minor, patch, pre), nil
+}
+
+func Latest(module string, includePre bool) error {
+	defer chdirRoot()()
+
+	_, latest, err := latestRelease(module, includePre)
 	if err != nil {
 		return err
 	}
 
-	tags := strings.Split(tagList, "\n")
-	semver.Sort(tags)
+	fmt.Println(latest)
 
-	lastTwo := tags[len(tags)-2:]
+	return nil
+}
 
-	// ensure we only attempt to publish updates when requested
-	// tag matches the latest semver in default branch.
-	if lastTwo[1] != tag {
-		return fmt.Errorf("Publish tag %q does not match latest semver tag %q. Aborting.", tag, lastTwo[1])
+func UpdateChangelog(module, version string) error {
+	defer chdirRoot()()
+
+	// If we're generating a pre-release then we compare to the last pre-release.
+	// Else, we compare with the latest non-prerelease to get maximum detail.
+	prefix, latest, err := latestRelease(module, semver.Prerelease(version) != "")
+	if err != nil {
+		return err
 	}
 
-	for _, submodule := range []string{
-		"errors",
-		"rpc/flipt",
-	} {
-		diff, err := base.WithExec([]string{"git", "diff", "--shortstat", lastTwo[0], lastTwo[1], "--", submodule}).Stdout(ctx)
-		if err != nil {
+	// go-git doesn't support half the functionality of git log
+	// It ends up brining along a lot of extra commits.
+	// So for now, I am just going to pop a shell.
+	rng := "HEAD"
+	if latest != "" {
+		if err := ensureVersionAfter(version, latest); err != nil {
 			return err
 		}
 
-		if strings.TrimSpace(diff) == "" {
-			fmt.Printf("Nothing changed between %q and %q. Skipping sub-module release.", lastTwo[0], lastTwo[1])
-			continue
-		}
+		rng = path.Join(prefix, latest) + ".." + rng
+	}
 
-		if err := tagSubmodule(ctx, ghClient, submodule, lastTwo[1]); err != nil {
-			return err
-		}
+	logCmd := fmt.Sprintf(`git --no-pager log --format="%%s" %s -- %s`, rng, prefix)
+	cmd := exec.Command("sh", "-c", logCmd)
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return err
+	}
+
+	messages := strings.Split(strings.TrimSpace(string(out)), "\n")
+
+	return insertChangeLogEntryIntoFile(
+		path.Join(prefix, "CHANGELOG.md"),
```

Comment:
could we use this to update the 'main' CHANGELOG for Flipt in the future?

Reply:
Indeed we can  🪄 :
```
cd hack/build
mage release:changelog go.flipt.io/flipt $(mage release:next go.flipt.io/flipt minor)
```

The output isn't perfect, but it certainly speeds up the process.

Reply:
`mage release:changelog go.flipt.io/flipt $(mage release:next go.flipt.io/flipt minor) | gpt4 > CHANGELOG.md` 💰